src/shell/task.c

```diff
@@ -0,0 +1,256 @@
+                          void *arg)
+{
+    task->io_cb = cb;
+    task->io_cb_arg = arg;
+    return 0;
+}
+
+/* N.B. Ref: cmd/flux-exec.c output_cb()
+ */
+int shell_task_io_readline (struct shell_task *task,
+                            const char *name,
+                            const char **line)
+{
+    const char *buf;
+    int len;
+
+    if (!(buf = flux_subprocess_read_line (task->proc, name, &len)))
+        return -1;
+    if (len == 0) {
+        if (flux_subprocess_state (task->proc) == FLUX_SUBPROCESS_EXITED) {
+            if (!(buf = flux_subprocess_read (task->proc, name, -1, &len)))
+                return -1;
+        }
+    }
+    *line = buf;
+    return len;
+}
+
+int shell_task_pmi_enable (struct shell_task *task,
+                           shell_task_pmi_ready_f cb,
+                           void *arg)
+{
+    if (flux_cmd_add_channel (task->cmd, "PMI_FD") < 0)
+        return -1;
+    if (flux_cmd_setenvf (task->cmd, 1, "PMI_RANK", "%d", task->rank) < 0)
+        return -1;
+    if (flux_cmd_setenvf (task->cmd, 1, "PMI_SIZE", "%d", task->size) < 0)
+        return -1;
+    task->pmi_cb = cb;
+    task->pmi_cb_arg = arg;
+    return 0;
+}
+
+int shell_task_pmi_readline (struct shell_task *task, const char **line)
+{
+    const char *buf;
+    int len;
+
+    if (!(buf = flux_subprocess_read_line (task->proc, "PMI_FD", &len)))
+        return -1;
+    *line = buf;
+    return len;
+}
+
+int shell_task_pmi_write (struct shell_task *task,
+                          const char *buf,
+                          int len)
+{
+    return flux_subprocess_write (task->proc, "PMI_FD", buf, len);
+}
+
+static void subproc_io_cb (flux_subprocess_t *p, const char *stream)
+{
+    struct shell_task *task = flux_subprocess_aux_get (p, "flux::task");
+
+    if (task->io_cb)
+        task->io_cb (task, stream, task->io_cb_arg);
+}
+
+static void subproc_channel_cb (flux_subprocess_t *p, const char *stream)
+{
+    struct shell_task *task = flux_subprocess_aux_get (p, "flux::task");
+
+    if (task->pmi_cb)
+        task->pmi_cb (task, task->pmi_cb_arg);
+}
+
+static void subproc_completion_cb (flux_subprocess_t *p)
+{
+    struct shell_task *task = flux_subprocess_aux_get (p, "flux::task");
+
+    if ((task->rc = flux_subprocess_exit_code (p)) < 0) {
+        if ((task->rc = flux_subprocess_signaled (p)) >= 0)
+            task->rc += 128;
+    }
+
+    if (task->cb)
+        task->cb (task, task->cb_arg);
+}
+
+static flux_subprocess_ops_t subproc_ops = {
+    .on_completion = subproc_completion_cb,
+    .on_state_change = NULL,
+    .on_channel_out = subproc_channel_cb,
+    .on_stdout = subproc_io_cb,
+    .on_stderr = subproc_io_cb,
+};
+
+
+int shell_task_start (struct shell_task *task,
+                      flux_reactor_t *r,
+                      shell_task_completion_f cb,
+                      void *arg)
+{
+    int flags = 0;
+
+    if (!task->io_cb)
+        flags |= FLUX_SUBPROCESS_FLAGS_STDIO_FALLTHROUGH;
```

Comment:
The intent was to (initially) fold I/O in with the shell's, for capture by the `job-exec` service, which dumps it to the flux log.  To prepare for a next step where I/O goes to your service, I added the callbacks and was intending, just for now, to have the callbacks implement the same behavior - emit to the shell's I/O, for capture by `job-exec`.

However when I did that, it seemed sometimes one task's output was lost in travis, so I left the little bit of code there but made it behave as before by adding the FALLTHROUGH flag.

I probably shouldn't have taken that list little step, but it may have uncovered some other problem with the shell terminating too soon, so it seems like a good thing to understand.

Reply:
I'm still trying to read and understand all the code. However, when you switched from "fallthrough" to having output handled by a callback, did you ensure to "flush" all the output from buffers of exited tasks. @chu11, correct me if I am wrong, but the "completed" callback for a subprocess only ensures EOF has been buffered on stdout/err for the process, not that it has been read or handled by a callback.

Reply:
I did not.  The code that failed is all there with the per-task initialization commented out.

Reply:
If something like the above isn't the problem, I wonder if the job-exec module could be responsible for dropping the output. Likely the bulk-exec code isn't super careful about handling all the IO. Maybe dump output to a file using the cb mechanism to see if job-exec is the culprit?

Reply:
Ah, just as a future note, I also noticed the callback will/might drop output if the tasks output doesn't end with a newline. I think flux-exec handles that case by checking in the output handler if a task has exited, and grabbing *all* remaining output if so.

Reply:
Oops, missed these comments yesterday.

>  @chu11, correct me if I am wrong, but the "completed" callback for a subprocess only ensures EOF has been buffered on stdout/err for the process, not that it has been read or handled by a callback.

Correct.  The EOF has been "sent" to the stdout/stderr callback, but there is no guarantee that it has been read.

Reply:
> Ah, just as a future note, I also noticed the callback will/might drop output if the tasks output doesn't end with a newline. I think flux-exec handles that case by checking in the output handler if a task has exited, and grabbing all remaining output if so.

I did copy that code from `flux-exec` (it is in shell/task.c):

```c
/* N.B. Ref: cmd/flux-exec.c output_cb()
 */
int shell_task_io_readline (struct shell_task *task,
                            const char *name,
                            const char **line)
{
    const char *buf;
    int len;

    if (!(buf = flux_subprocess_read_line (task->proc, name, &len)))
        return -1;
    if (len == 0) {
        if (flux_subprocess_state (task->proc) == FLUX_SUBPROCESS_EXITED) {
            if (!(buf = flux_subprocess_read (task->proc, name, -1, &len)))
                return -1;
        }
    }
    *line = buf;
    return len;
}
```

I'm not sure it matters, but I am not synchronizing using the "completed" callback, but instead just letting the reactor exit naturally when subprocess unregisters all its watchers.

I have a commit waiting to try (after verifying valgrind issue is gone) that will call the I/O callback one last time after the reactor exits to consume any buffered I/O, although I'm pretty sure I tried that before and it didn't help.

Reply:
Ah ok, sorry I missed that. Since you have that I'm wondering again if the problem is in the job-exec module.

BTW, how were you detecting dropped lines of output?

Reply:
`t2601-job-shell-standalone.t` runs the job shell without `job-exec` and the missing lines of output are detected there, so `job-exec` is innocent!  Sorry, should have spoken up when you mentioned that before.

Reply:
> t2601-job-shell-standalone.t runs the job shell without job-exec and the missing lines of output are detected there

Hm, I uncommented `shell_task_io_enable` and don't see an error in that test
```diff
diff --git a/src/shell/shell.c b/src/shell/shell.c
index db54b19..c46e77a 100644
--- a/src/shell/shell.c
+++ b/src/shell/shell.c
@@ -209,8 +209,8 @@ int main (int argc, char *argv[])
             log_err_exit ("shell_task_create index=%d", i);
         if (shell_task_pmi_enable (task, shell_pmi_task_ready, pmi) < 0)
             log_err_exit ("shell_task_pmi_enable");
-        //if (shell_task_io_enable (task, shell_io_task_ready, io) < 0)
-        //    log_err_exit ("shell_task_io_enable");
+        if (shell_task_io_enable (task, shell_io_task_ready, io) < 0)
+            log_err_exit ("shell_task_io_enable");
         if (shell_task_start (task, r, task_completion_cb, info) < 0)
             log_err_exit ("shell_task_start index=%d", i);
```

```
~/git/flux-core.git/t$ ./t2601-job-shell-standalone.t 
ok 1 - flux-shell: generate 1-task jobspec and matching R
ok 2 - flux-shell: run 1-task echo job
ok 3 - flux-shell: -v output includes expected task count
ok 4 - flux-shell: missing JOBID fails with Usage message
ok 5 - flux-shell: missing -r fails with standalone message
ok 6 - flux-shell: missing -R fails with standalone message
ok 7 - flux-shell: missing -j fails with standalone message
ok 8 - flux-shell: nonexistent jobspec file fails
ok 9 - flux-shell: malformed jobid fails
ok 10 - flux-shell: out of range jobid fails
ok 11 - flux-shell: out of range broker rank fails
ok 12 - flux-shell: wrong range broker rank fails
ok 13 - flux-shell: unknown argument fails
ok 14 - flux-shell: generate 2-task jobspec and matching R
ok 15 - flux-shell: run 2-task printenv job
ok 16 - flux-shell: 0: completed with no error
ok 17 - flux-shell: 1: completed with no error
ok 18 - flux-shell: 0: FLUX_TASK_LOCAL_ID, TASK_RANK set
ok 19 - flux-shell: 1: FLUX_TASK_LOCAL_ID, TASK_RANK set
ok 20 - flux-shell: FLUX_JOB_SIZE, JOB_NNODES, JOB_ID set
ok 21 - flux-shell: 0: PMI_RANK set
ok 22 - flux-shell: 1: PMI_RANK set
ok 23 - flux-shell: PMI_SIZE, PMI_FD set
ok 24 - flux-shell: generate 8-task bash exit rank job
ok 25 - flux-shell: environ in jobspec is set for task
ok 26 - flux-shell: shell PMI works
ok 27 - flux-shell: shell PMI KVS works
ok 28 - flux-shell: shell can launch flux
ok 29 - flux-shell: shell exits with highest task exit value
# passed all 29 test(s)
1..29
```

Reply:
I wasn't able to reproduce as well.

Reply:
I've only seen it in travis, sadly.

Reply:
Yeah, I think I see it in the current travis run. Perhaps the constrained travis container causes a race to be triggered. (though it definitely isn't obvious what the problem could be as all the code looks fine to me.)

Reply:
I've found I can reproduce by binding a shell in standalone mode to a single core:
```
$ taskset -c 0 ../src/shell/flux-shell -s -v -r 0 -R trash-directory.t2601-job-shell-standalone/R2 -j trash-directory.t2601-job-shell-standalone/j2 123456 > j2.out 
flux-shell: 0: task_count=2 slot_count=2 cores_per_slot=1 slots_per_node=2
flux-shell: 0: tasks [0-1] on cores 0-1
0: C: pmi EOF
flux-shell: task 0 complete status=0
1: C: pmi EOF
flux-shell: task 1 complete status=0
flux-shell: exit 0
$ grep -c ^0: j2.out
109
$ grep -c ^1: j2.out
36
```

(similarly, running `t2601-job-shell-standalone.t` under `taskset -c 0` reproduces the travis failures) Maybe that will help find root cause?

Reply:
Oooh, nice!   I can't seem to reproduce it on my virtualbox vm with that (and anyway it's only got one core).  Maybe I need to fire up some different hardware here to get it to happen.

Reply:
I added some printf debug to `shell_task_io_readline` and the results are a bit strange. 

```diff
diff --git a/src/shell/task.c b/src/shell/task.c
index 16b48e3..06e32de 100644
--- a/src/shell/task.c
+++ b/src/shell/task.c
@@ -148,10 +151,12 @@ int shell_task_io_readline (struct shell_task *task,
 
     if (!(buf = flux_subprocess_read_line (task->proc, name, &len)))
         return -1;
+    fprintf (stderr, "%d: %s: read %d bytes\n", task->rank, name, len);
     if (len == 0) {
         if (flux_subprocess_state (task->proc) == FLUX_SUBPROCESS_EXITED) {
             if (!(buf = flux_subprocess_read (task->proc, name, -1, &len)))
                 return -1;
+            fprintf (stderr, "%d: %s: after exit: %d bytes\n", task->rank, name, len);
         }
     }
     *line = buf;
```

This might not be directly related to the case of the missing lines though

```
[snip a bunch of 0: STDOUT lines]

0: STDOUT: read 16 bytes
0: STDOUT: read 18 bytes
0: STDOUT: read 19 bytes
0: STDOUT: read 11 bytes
0: STDOUT: read 11 bytes
0: STDOUT: read 10 bytes
0: STDOUT: read 0 bytes
0: STDOUT: after exit: 0 bytes
0: STDOUT: read 0 bytes
0: STDOUT: after exit: 0 bytes
0: STDERR: read 0 bytes
0: STDERR: after exit: 0 bytes
flux-shell: task 0 complete status=0
1: STDOUT: read 19 bytes
1: STDOUT: read 9 bytes
1: STDOUT: read 85 bytes
1: STDOUT: read 12 bytes
1: STDOUT: read 1421 bytes
1: STDOUT: read 15 bytes
1: STDOUT: read 122 bytes
1: STDOUT: read 15 bytes
1: STDOUT: read 25 bytes
1: STDOUT: read 38 bytes
1: STDOUT: read 45 bytes
1: STDOUT: read 15 bytes
1: STDOUT: read 330 bytes
1: STDOUT: read 218 bytes
1: STDOUT: read 11 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes

// above repeats about 3000x
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 1 bytes
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 0 bytes
1: C: pmi EOF
1: STDERR: read 0 bytes
1: STDERR: after exit: 0 bytes
flux-shell: task 1 complete status=0
0: STDOUT: read 0 bytes
0: STDOUT: after exit: 0 bytes
0: STDERR: read 0 bytes
0: STDERR: after exit: 0 bytes
0: STDOUT: read 0 bytes
0: STDOUT: after exit: 0 bytes
0: STDERR: read 0 bytes
0: STDERR: after exit: 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 0 bytes
1: STDERR: read 0 bytes
1: STDERR: after exit: 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 0 bytes
1: STDERR: read 0 bytes
1: STDERR: after exit: 0 bytes
flux-shell: exit 0
```

I'm not sure this tells us that much. However a couple points:

 - It will save a lot of useless looping if line buffered streams aren't "ready" for reading unless at least one line is available to read.

 - I'm not sure how EOF handling works with local subprocess buffered I/O. I don't see where an EOF is marked as pending or consumed on streams (obviously this must work somehow, or have we just not dealt with local EOFs yet?)

Other than that, it looks like task 1 buffer just stops reading data after some number of lines. Not sure why yet. (sorry for the inconclusive data post, but figured maybe @chu11 might have some ideas where to look next)

Reply:
Another hint, added a printf line in `subprocess_read` to display the size of the buffer we're reading from and on a failing run I see, for task 0:
```
subprocess_read: STDOUT: lines=109 bytes=8767
0: STDOUT: read 36 bytes
subprocess_read: STDOUT: lines=108 bytes=8731
0: STDOUT: read 278 bytes
...
```

whereas task 1 shows:

```
flux-shell: task 0 complete status=0
subprocess_read: STDOUT: lines=35 bytes=4096
1: STDOUT: read 36 bytes
subprocess_read: STDOUT: lines=34 bytes=4060
1: STDOUT: read 278 bytes
...
flux-shell: task 0 complete status=0
subprocess_read: STDOUT: lines=35 bytes=4096
1: STDOUT: read 36 bytes
subprocess_read: STDOUT: lines=34 bytes=4060
1: STDOUT: read 278 bytes
```

Suspciously task 1 buffer is only filled to 4K bytes and never gets any more data...

Reply:
> It will save a lot of useless looping if line buffered streams aren't "ready" for reading unless at least one line is available to read.

This seems like a useful option to have for `libsubprocess`.  Right now the watcher is not line buffered.  Should be easy to add with `FLUX_WATCHER_LINE_BUFFER`.

> 1: STDOUT: read 0 bytes
> 1: STDOUT: read 0 bytes
> 1: STDOUT: read 0 bytes

my initial thought was if there was some bug where the watcher was not shut down properly and constantly calling the `on_stdout` callback.  But it appears that is shut off once EOF is sent to the caller.

Are you reproducing this via the `taskset` reproducer you have above?

Reply:
> Are you reproducing this via the taskset reproducer you have above?

Yeah, the simplest reproducer is to first run `./t2601-job-shell-standalone -d -v` to get the appropriate input files in `trash-directory.t2601-job-shell-standalone`, then run:

```
 taskset -c 0 ../src/shell/flux-shell -s -v -r 0 -R trash-directory.t2601-job-shell-standalone/R2 -j trash-directory.t2601-job-shell-standalone/j2 123456 > j2.out 2>log
```

On fluke, this seems to reproduce the issue 100%.

I'm actually wondering if we should try to make a more isolated reproducer that just runs two local subprocesses to take the job-shell out of the equation. Something funky seems to be going on with buffered io here. In my testcase I added some printf debug down in the ev_buffer_read code and I'm not seeing where the subprocess buffers are even reading EOF from their stdout streams...

```console
$ taskset -c 0 ../src/shell/flux-shell -s -v -r 0 -R trash-directory.t2601-job-shell-standalone/R2 -j trash-directory.t2601-job-shell-standalone/j2 123456 > j2.out 2>log
$ grep buffer_read_cb log
buffer_read_cb: fd=13, space=1048576, read=0 bytes
buffer_read_cb: fd=13. eof_read=true
buffer_read_cb: fd=9, space=1048576, read=8767 bytes
buffer_read_cb: fd=11, space=1048576, read=0 bytes
buffer_read_cb: fd=11. eof_read=true
buffer_read_cb: fd=12, space=1048576, read=4096 bytes
buffer_read_cb: fd=17, space=1048576, read=0 bytes
buffer_read_cb: fd=17. eof_read=true
buffer_read_cb: fd=15, space=1048576, read=0 bytes
buffer_read_cb: fd=15. eof_read=true
```

Note no `eof_read=true` for fds 9 and 12, the stdout fds of task 0 and 1 respectively.

Reply:
Note I see the same issue with no eof on the non-failing reproducer (without `taskset -c 0`), however the difference is that the initial `read` into the buffer for task 1 is the full `8767` bytes.

Makes me wonder if we make a test to output much larger files if data will be dropped even in the non `taskset -c0` case.

Reply:
Ok, got it reproduced (took me awhile to realize, oh I should probably be in a flux instance).  Reproducing on catalyst and fluke consistently.  Can dig into this deeper tomorrow.

Reply:
So I added some debugging to the buffer reactor (libflux/ev_buffer_read.c)

> buffer_read_cb:96 write=4096
> subprocess_read: STDOUT: lines=89 bytes=4096
> 1: STDOUT: read 26 bytes

AFAICT, the task is only writing 4096 bytes, so apparently only 4096 bytes of data are coming
from task 1.  So that's an interesting nugget.

Reply:
> AFAICT, the task is only writing 4096 bytes, so apparently only 4096 bytes of data are coming from task 1.

I think it unlikely that task 1 is only writing 4096 bytes. I think I did try running something other than `printenv` which had guaranteed same output from each task and still some of task 1 stdout was dropped.

Reply:
Sorry, what I meant was "AFAICT, the task is only writing 4096 bytes instead of the (> 4096) bytes it should be."  

I forced libsubprocess to be line buffered and the reproducer started working.  Saw:

> buffer_read_cb:96 write=4096
> buffer_read_cb:96 write=247

in task 1, which is equal to task 0's

> buffer_read_cb:96 write=4343

I suppose task 1 isn't `fflush`-ing or something like task 0, but somehow on the second write
isn't completing.

Reply:
Ah, good hint! However, shouldn't a process exiting force stdio flush?

Reply:
> However, shouldn't a process exiting force stdio flush?

I would think so.

Reply:
Looks like an `atexit()` handler flushes buffered stdio when `exit()` is called:
https://www.gnu.org/software/libc/manual/html_node/Flushing-Buffers.html

Reply:
so I think I have a lead on the problem.  There seems to be a circumstance where the `flux_buffer_read_watcher()` does NOT read EOF on a stream, BUT `libsubprocess` calculates that EOF has occurred and sends EOF to a `libsubprocess` callback. 

I'm zero-ing on this part of the code in libsubprocess in `local.c`.  I'm thinking that the `output_cb` (which calls the user's `on_stdout`) reads all the data in the buffer.  The child process has exited, BUT all of the stdout hasn't been flushed / read into the buffer yet.  So this is not a safe calculation of EOF.  Will have to look into this more tomorrow and figure out a reproducer, tests, and how to fix.  And if the problem exists for non-local subprocesses too.

I hope this is it, this is a bit of a doozy :-)

```
        output_cb (c->p, c->name);

        if (c->p->state == FLUX_SUBPROCESS_EXITED && !c->eof_sent_to_caller) {

            if (!(fb = flux_buffer_read_watcher_get_buffer (w))) {
                flux_log_error (c->p->h, "flux_buffer_read_watcher_get_buffer");
                return;
            }

            if (!flux_buffer_bytes (fb)) {

                output_cb (c->p, c->name);

                c->eof_sent_to_caller = true;
                eof_set = true;
                c->p->channels_eof_sent++;
            }
        }

```

Reply:
Nice sleuthing! That looks promising to me! We could probably reproduce by running a local subprocess with a lot of output (so that child exit is detected before all output is complete), and maybe shrinking the childs stdio buffer size (i.e. via setvbuf()) might make this situation more likely to occur. 

BTW, should we move this to its own issue (and copy the last few comments here)?

Reply:
Tomorrow I'll try to 100% confirm it.  Many many random printf's all over the place, want to try and confirm it is specifically this and that `ev_io()` and the `flux_buffer_read_watcher()` work as I believe it should.

Reply:
is now issue #2228